### PR TITLE
optimize(core): move internal js src off heap

### PR DESCRIPTION
**Pull request by AaronO**
This PR explores moving the internal JS code (bundled in `deno`) off the heap, as suggested by https://github.com/denoland/deno/issues/9675

## Impact

Reduce per-isolate heap by `~1MB`, reducing overall memory usage (especially with many workers), with no real downside or compromise.

This should also reduce snapshot sizes, resulting in faster snapshot restores and a smaller `deno` binary (these will likely be minor and weren't the core motivation)

## Subtasks

- [ ] ASCII JS code
  - [x] Ensure internal JS files are ASCII
  - [ ] Add a lint / CI test check to ensure files remain ASCII
- [x] Allow executing JS code via external `v8::String`s
  - [x] Implement external strings in `rusty_v8` (https://github.com/denoland/rusty_v8/pull/641)
  - [x] Implement `JsRuntime::execute_static`
- [ ] Work with snapshots
  - [ ] Implement support for string `ExternalReferences` in v8, so they can be captured and restored in snapshots
  - [ ] Register internal src references with the snapshot creator (maybe as they're passed through `execute_static` ?)

## Notes

Non-ASCII files found with (ignores test and example files):
```
find core runtime op_crates -name "*.[jt]s" | grep -v "examples" | grep -v "_test" | xargs file | grep -v ASCII
```

Non-ASCII characters in each file were then found with this regex: `[^\x00-\x7F]`

Non-ASCII files and context:
- `runtime/js/02_console.js`: the non-ASCII chars are the tableChars (L38)
- `op_crates/web/02_event.js`: there's a single utf8 apostrophe `’` vs `'` in comments (L919)
- `op_crates/web/21_filereader.js`: there are multiple utf8 apostrophes in comments (same as above)
- `op_crates/fetch/26_fetch.js`: same deal, same apostrophes
- `op_crates/web/lib.deno_web.d.ts`: used `§` in comment

### CODE REVIEW

**Review by ry — COMMENTED** [review]
Awesome! This is a great simple optimization

File: op_crates/web/02_event.js [hunk 1]
```diff
@@ -916,7 +916,7 @@
           // If signal is not null and its aborted flag is set, then return.
           return;
         } else {
-          // If listener’s signal is not null, then add the following abort
```
Comment (ry) [inline op_crates/web/02_event.js:919]: This character non-ascii?

Comment (caspervonb) [reply op_crates/web/02_event.js:919]: Really subtle but this is a *right single quotation mark* (0x2019).
A quick `s/’/'/g` and we should be good.

**Comment by ry** [comment]
https://github.com/denoland/deno/pull/9725

**Comment by AaronO** [comment]
Thanks @ry. Btw, this will still require some further changes in `rusty_v8`, specifically for:
>  Implement support for string ExternalReferences in v8, so they can be captured and restored in snapshots

So external strings work, just not with snapshots yet. So I wouldn't bump `rusty_v8` solely for this PR without the `ExternalReference` changes.

Locally I've made the necessary changes to register the external refs (beyond just `bindings::EXTERNAL_REFERENCES`) and pass in these string pointers, everything runs but it doesn't impact the heap-usage as intended.

So this will require a little more investigating on my end.

**Comment by jbergstroem** [comment]
Perhaps there should be a CI job to check for this in future code changes. Add to linting?

**Comment by AaronO** [comment]
@jbergstroem It's planned, see the subtask:
>  Add a lint / CI test check to ensure files remain ASCII

**Comment by bartlomieju** [comment]
I will add relevant bindings to `rusty_v8` to support this

**Comment by magurotuna** [comment]
Completed to implement a new lint rule in https://github.com/denoland/deno_lint/pull/837.
Quickly running the rule on `ext`, `runtime` and `core`, I've got the following errors.

## ext

Found 22 problems. I'm worried that non-ASCII characters used in `02_console.js` are required to implement `console.table`.

<details>
<summary>click to expand</summary>

```
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:129:20
    |
129 |     middleMiddle: "─",
    |                    ^
    |
    = help: `─` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:130:17
    |
130 |     rowMiddle: "┼",
    |                 ^
    |
    = help: `┼` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:131:16
    |
131 |     topRight: "┐",
    |                ^
    |
    = help: `┐` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:132:15
    |
132 |     topLeft: "┌",
    |               ^
    |
    = help: `┌` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:133:18
    |
133 |     leftMiddle: "├",
    |                  ^
    |
    = help: `├` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:134:17
    |
134 |     topMiddle: "┬",
    |                 ^
    |
    = help: `┬` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:135:19
    |
135 |     bottomRight: "┘",
    |                   ^
    |
    = help: `┘` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:136:18
    |
136 |     bottomLeft: "└",
    |                  ^
    |
    = help: `└` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:137:20
    |
137 |     bottomMiddle: "┴",
    |                    ^
    |
    = help: `┴` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:138:19
    |
138 |     rightMiddle: "┤",
    |                   ^
    |
    = help: `┤` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:139:12
    |
139 |     left: "│ ",
    |            ^
    |
    = help: `│` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:140:14
    |
140 |     right: " │",
    |              ^
    |
    = help: `│` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/console/02_console.js:141:15
    |
141 |     middle: " │ ",
    |               ^
    |
    = help: `│` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/web/02_event.js:927:25
    |
927 |           // If listener’s signal is not null, then add the following abort
    |                         ^
    |
    = help: `’` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/web/10_filereader.js:64:18
   |
64 |       // 1. If fr’s state is "loading", throw an InvalidStateError DOMException.
   |                  ^
   |
   = help: `’` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/web/10_filereader.js:71:19
   |
71 |       // 2. Set fr’s state to "loading".
   |                   ^
   |
   = help: `’` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/web/10_filereader.js:73:19
   |
73 |       // 3. Set fr’s result to null.
   |                   ^
   |
   = help: `’` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/web/10_filereader.js:75:19
   |
75 |       // 4. Set fr’s error to null.
   |                   ^
   |
   = help: `’` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/web/10_filereader.js:149:29
    |
149 |                 // 1. Set fr’s state to "done".
    |                             ^
    |
    = help: `’` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/web/10_filereader.js:151:87
    |
151 |                 // 2. Let result be the result of package data given bytes, type, blob’s type, and encodingName.
    |                                                                                       ^
    |
    = help: `’` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/web/10_filereader.js:226:28
    |
226 |                 // 5. If fr’s state is not "loading", fire a progress event called loadend at the fr.
    |                            ^
    |
    = help: `’` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
   --> /Users/yusuktan/Repos/github.com/magurotuna/deno/ext/web/10_filereader.js:253:22
    |
253 |               //If fr’s state is not "loading", fire a progress event called loadend at fr.
    |                      ^
    |
    = help: `’` is not an ASCII. Consider replacing it
Found 22 problems
```

</details>

## runtime

No non-ASCII characters are found.

## core 

Found 10 problems, but all of them are in test file `encode_decode_test.js`

<details>
<summary>click to expand</summary>

```
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/core/encode_decode_test.js:39:50
   |
39 |   assertArrayEquals(Array.from(Deno.core.encode("𝓽𝓮𝔁𝓽")), fixture1);
   |                                                  ^
   |
   = help: `𝓽` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/core/encode_decode_test.js:39:51
   |
39 |   assertArrayEquals(Array.from(Deno.core.encode("𝓽𝓮𝔁𝓽")), fixture1);
   |                                                   ^
   |
   = help: `𝓮` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/core/encode_decode_test.js:39:52
   |
39 |   assertArrayEquals(Array.from(Deno.core.encode("𝓽𝓮𝔁𝓽")), fixture1);
   |                                                    ^
   |
   = help: `𝔁` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/core/encode_decode_test.js:39:53
   |
39 |   assertArrayEquals(Array.from(Deno.core.encode("𝓽𝓮𝔁𝓽")), fixture1);
   |                                                     ^
   |
   = help: `𝓽` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/core/encode_decode_test.js:48:58
   |
48 |   assert(Deno.core.decode(new Uint8Array(fixture1)) === "𝓽𝓮𝔁𝓽");
   |                                                          ^
   |
   = help: `𝓽` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/core/encode_decode_test.js:48:59
   |
48 |   assert(Deno.core.decode(new Uint8Array(fixture1)) === "𝓽𝓮𝔁𝓽");
   |                                                           ^
   |
   = help: `𝓮` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/core/encode_decode_test.js:48:60
   |
48 |   assert(Deno.core.decode(new Uint8Array(fixture1)) === "𝓽𝓮𝔁𝓽");
   |                                                            ^
   |
   = help: `𝔁` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/core/encode_decode_test.js:48:61
   |
48 |   assert(Deno.core.decode(new Uint8Array(fixture1)) === "𝓽𝓮𝔁𝓽");
   |                                                             ^
   |
   = help: `𝓽` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/core/encode_decode_test.js:49:64
   |
49 |   assert(Deno.core.decode(new Uint8Array(fixture2)) === "Hello �� World");
   |                                                                ^
   |
   = help: `�` is not an ASCII. Consider replacing it
error[prefer-ascii]: Non-ASCII characters are not allowed
  --> /Users/yusuktan/Repos/github.com/magurotuna/deno/core/encode_decode_test.js:49:65
   |
49 |   assert(Deno.core.decode(new Uint8Array(fixture2)) === "Hello �� World");
   |                                                                 ^
   |
   = help: `�` is not an ASCII. Consider replacing it
Found 10 problems
```

</details>

**Comment by AaronO** [comment]
@magurotuna Great job !

A nice additional improvement would be to show the escaped `\x...` value

**Comment by ry** [comment]
Closing because stale. Open a new PR if this work continues.

**Pull request closed without merging**